File: backend/src/main/java/com/telehealth/pro/controller/NotificationController.java
package com.telehealth.pro.controller;

import com.telehealth.pro.dto.ApiResponse;
import com.telehealth.pro.entity.Notification;
import com.telehealth.pro.entity.User;
import com.telehealth.pro.repository.NotificationRepository;
import com.telehealth.pro.security.CustomUserDetails;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/notifications")
@PreAuthorize("isAuthenticated()")
public class NotificationController {

    @Autowired
    private NotificationRepository notificationRepository;

    @GetMapping
    public ResponseEntity<?> listNotifications(@AuthenticationPrincipal CustomUserDetails userDetails) {
        User currentUser = userDetails.getUser();
        List<Notification> notifications = notificationRepository
                .findAllByUser_IdOrderByCreatedAtDesc(currentUser.getId());
        return ResponseEntity.ok(ApiResponse.success(notifications));
    }

    @PutMapping("/{id}/read")
    @Transactional
    public ResponseEntity<?> markAsRead(
            @AuthenticationPrincipal CustomUserDetails userDetails,
            @PathVariable String id) {
        
        User currentUser = userDetails.getUser();

        if ("all".equalsIgnoreCase(id)) {
            notificationRepository.markAllAsReadForUser(currentUser.getId());
            return ResponseEntity.ok(ApiResponse.builder()
                    .success(true)
                    .message("All notifications marked as read.")
                    .build());
        }

        try {
            Long notificationId = Long.parseLong(id);
            Optional<Notification> notificationOpt = notificationRepository
                    .findByNotificationIdAndUser_Id(notificationId, currentUser.getId());
            
            if (notificationOpt.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(ApiResponse.error("Notification not found."));
            }

            Notification notification = notificationOpt.get();
            notification.setRead(true);
            notification = notificationRepository.save(notification);

            return ResponseEntity.ok(ApiResponse.success("Notification marked as read.", notification));
        } catch (NumberFormatException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                    .body(ApiResponse.error("Invalid notification ID format."));
        }
    }
}

File: backend/src/controllers/notification.controller.js
const { Notification } = require('../models');

const listNotifications = async (req, res) => {
  try {
    const notifications = await Notification.findAll({
      where: { user_id: req.user.id },
      order: [['created_at', 'DESC']]
    });

    res.status(200).json({ success: true, data: notifications });
  } catch (error) {
    console.error('List Notifications Error:', error);
    res.status(500).json({ success: false, message: 'Failed to retrieve notifications.' });
  }
};

const markAsRead = async (req, res) => {
  try {
    const { id } = req.params;

    if (id === 'all') {
      await Notification.update(
        { is_read: true },
        { where: { user_id: req.user.id } }
      );
      return res.status(200).json({ success: true, message: 'All notifications marked as read.' });
    }

    const notification = await Notification.findOne({
      where: { notification_id: id, user_id: req.user.id }
    });

    if (!notification) {
      return res.status(404).json({ success: false, message: 'Notification not found.' });
    }

    notification.is_read = true;
    await notification.save();

    res.status(200).json({ success: true, message: 'Notification marked as read.', data: notification });
  } catch (error) {
    console.error('Mark Notification Read Error:', error);
    res.status(500).json({ success: false, message: 'Failed to update notification state.' });
  }
};

module.exports = {
  listNotifications,
  markAsRead
};
